cineast-core/src/main/java/org/vitrivr/cineast/core/features/InceptionResnetV2.java

```diff
@@ -0,0 +1,205 @@
+
+      if (image != null) {
+        float[] encodingArray = encodeImage(image);
+        this.persist(shot.getId(), new FloatVectorImpl(encodingArray));
+      }
+
+      // Insert return here if additional cases are added!
+    }
+  }
+
+  @Override
+  public List<ScoreElement> getSimilar(SegmentContainer sc, ReadableQueryConfig qc) {
+    if (sc.getMostRepresentativeFrame() == VideoFrame.EMPTY_VIDEO_FRAME) {
+      LOGGER.error("Could not get similar because no image was provided.");
+      return new ArrayList<>();
+    }
+
+    BufferedImage image = sc.getMostRepresentativeFrame().getImage().getBufferedImage();
+
+    if (image == null) {
+      LOGGER.error("Could not get similar because image could not be converted to BufferedImage.");
+      return new ArrayList<>();
+    }
+
+    // Ensure the correct distance function is used
+    QueryConfig queryConfig = QueryConfig.clone(qc);
+    queryConfig.setDistance(DISTANCE);
+
+    float[] encodingArray = encodeImage(image);
+
+    return getSimilar(encodingArray, queryConfig);
+  }
+
+  @Override
+  public List<ScoreElement> getSimilar(String segmentId, ReadableQueryConfig qc) {
+    // Ensure the correct distance function is used
+    QueryConfig queryConfig = QueryConfig.clone(qc);
+    queryConfig.setDistance(DISTANCE);
+
+    return super.getSimilar(segmentId, queryConfig);
+  }
+
+  public static SavedModelBundle getModel() {
+    initializeModel();
+
+    return model;
+  }
+
+  /**
+   * Encodes the given image using InceptionResnetV2.
+   *
+   * @return Intermediary encoding, not yet embedded.
+   */
+  public static float[] encodeImage(BufferedImage image) {
+    initializeModel();
+
+    if (image.getWidth() != IMAGE_WIDTH || image.getHeight() != IMAGE_HEIGHT) {
+      image = rescale(image, IMAGE_WIDTH, IMAGE_HEIGHT);
+    }
+    int[] colors = image.getRGB(0, 0, IMAGE_WIDTH, IMAGE_HEIGHT, null, 0, IMAGE_WIDTH);
+    int[] rgb = colorsToRGB(colors);
+    float[] processedColors = preprocessInput(rgb);
+
+    try (TFloat32 imageTensor = TFloat32.tensorOf(Shape.of(1, IMAGE_WIDTH, IMAGE_HEIGHT, 3), DataBuffers.of(processedColors))) {
+      HashMap<String, Tensor> inputMap = new HashMap<>();
+      inputMap.put(INPUT, imageTensor);
+
+      Map<String, Tensor> resultMap = model.call(inputMap);
+
+      try (TFloat32 encoding = (TFloat32) resultMap.get(OUTPUT)) {
+
+        float[] embeddingArray = new float[ENCODING_SIZE];
+        FloatDataBuffer floatBuffer = DataBuffers.of(embeddingArray);
+        encoding.read(floatBuffer);
+
+        return embeddingArray;
+      }
+    }
+  }
+
+  // TODO: Move image util related functions to a dedicated util class
+
+  /**
+   * Rescales a buffered image using bilinear interpolation.
+   */
+  public static BufferedImage rescale(BufferedImage image, int width, int height) {
+    BufferedImage scaledImage = new BufferedImage(width, height, image.getType());
```

Comment:
We usually use another library for image operations, since `awt` might (in some rare instances) not be available on headless machines. Currently, the `Thumbnailator` is used (although I might want to replace that at some point since there are other libraries that are more actively maintained).

Reply:
I'm not sure I understand your concern correctly; I had a quick look at Thumbnailator and it is also [depending on awt](https://github.com/coobird/thumbnailator/blob/b97f8fba767a5a0b5378bdd9cb3c81f39b295b00/src/main/java/net/coobird/thumbnailator/Thumbnailator.java#L27).

Reply:
One more reason to replace it then, although the dependency might only be for compatibility to the data structures, I'd have to check. The main point is: there is a more readable and potentially more efficient way to achieve the same result which is (ideally) consistently used by all the other features, so I don't see a reason not to use it here as well.

Reply:
Unfortunately, I have discovered that the `Thumbnailator` can only resize images while preserving the original aspect ratio, something that is not compatible with the input requirements of InceptionResnetV2. Since the replacement of the `Thumbnailator` is an entirely different issue to this pull request, I suggest merging this as-is and opening a new issue for the replacement and unification of image utilities.

Reply:
Nope, if you want to change the aspect ratio, simply use `forceSize(w, h)` or `size(w, h).keepAspectRatio(false)` as stated in the documentation. I also just saw that the maintenance appears to have picked up, so there's probably no need to change it after all. An update might be in order though.

Reply:
My bad, I had misunderstood the documentation.